File: ios/HindPharmaApp/ViewModel/HomeViewModel.swift
import Combine
import Foundation

@MainActor
final class HomeViewModel: ObservableObject {
    @Published var username = ""
    @Published var password = ""
    @Published private(set) var sessionUser: SessionUser?
    @Published private(set) var isLoading = false
    @Published var errorMessage: String?

    func login() async {
        let cleanUsername = username.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cleanUsername.isEmpty else {
            errorMessage = "Username is required."
            return
        }
        guard !password.isEmpty else {
            errorMessage = "Password is required."
            return
        }

        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        do {
            var request = URLRequest(url: AppConfig.apiBaseURL.appendingPathComponent("login"))
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: ["username": cleanUsername, "password": password])

            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
                let detail = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["detail"] as? String
                throw LoginError.server(detail ?? "Invalid username or password")
            }

            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let responseUsername = json["username"] as? String,
                  let role = json["role"] as? String,
                  let token = json["token"] as? String else {
                throw LoginError.server("Invalid login response")
            }

            sessionUser = SessionUser(
                id: String(describing: json["id"] ?? ""),
                username: responseUsername,
                role: role,
                adminID: json["admin_id"].map { String(describing: $0) },
                businessName: json["business_name"] as? String,
                subscriptionExpiry: json["subscription_expiry"] as? String,
                token: token
            )
            password = ""
        } catch {
            sessionUser = nil
            errorMessage = error.localizedDescription
        }
    }

    func logout() {
        sessionUser = nil
        username = ""
        password = ""
        errorMessage = nil
    }

    private enum LoginError: LocalizedError {
        case server(String)
        var errorDescription: String? {
            switch self {
            case .server(let message): return message
            }
        }
    }
}

File: android/app/src/main/java/com/hindtechgroup/hindpharma/ViewModel/HomeViewModel.kt
package com.hindtechgroup.hindpharma.ViewModel

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import com.hindtechgroup.hindpharma.Helper.AppConfig
import com.hindtechgroup.hindpharma.Models.SessionUser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

class HomeViewModel : ViewModel() {
    var username by mutableStateOf("")
    var password by mutableStateOf("")
    var isLoading by mutableStateOf(false)
    var errorMessage by mutableStateOf<String?>(null)
    var sessionUser by mutableStateOf<SessionUser?>(null)
        private set

    suspend fun login() {
        val cleanUsername = username.trim()
        if (cleanUsername.isEmpty()) {
            errorMessage = "Username is required."
            return
        }
        if (password.isEmpty()) {
            errorMessage = "Password is required."
            return
        }

        isLoading = true
        errorMessage = null
        try {
            sessionUser = withContext(Dispatchers.IO) {
                val connection = (URL("${AppConfig.apiBaseUrl}/login").openConnection() as HttpURLConnection)
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.connectTimeout = 10_000
                connection.readTimeout = 10_000
                connection.doOutput = true
                connection.outputStream.use {
                    it.write(JSONObject().apply {
                        put("username", cleanUsername)
                        put("password", password)
                    }.toString().toByteArray())
                }

                val responseCode = connection.responseCode
                val stream = if (responseCode in 200..299) connection.inputStream else connection.errorStream
                val body = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
                connection.disconnect()

                if (responseCode !in 200..299) {
                    val message = runCatching { JSONObject(body).optString("detail") }.getOrNull()
                    throw IllegalStateException(message?.ifBlank { null } ?: "Invalid username or password")
                }

                val json = JSONObject(body)
                SessionUser(
                    id = json.optString("id"),
                    username = json.optString("username"),
                    role = json.optString("role", "employee"),
                    adminId = if (json.isNull("admin_id")) null else json.optString("admin_id"),
                    businessName = if (json.isNull("business_name")) null else json.optString("business_name"),
                    subscriptionExpiry = if (json.isNull("subscription_expiry")) null else json.optString("subscription_expiry"),
                    token = json.optString("token")
                )
            }
            password = ""
        } catch (error: Exception) {
            errorMessage = error.message ?: "Unable to complete login."
            sessionUser = null
        } finally {
            isLoading = false
        }
    }

    fun logout() {
        sessionUser = null
        username = ""
        password = ""
        errorMessage = null
    }
}
